consensus/protocol_upgrade_table.go

```diff
@@ -0,0 +1,68 @@
+package consensus
+
+import (
+	"sort"
+
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// protocolUpgrade specifies that a particular protocol version goes into effect at a particular block height
+type protocolUpgrade struct {
+	Version     uint64
+	EffectiveAt *types.BlockHeight
+}
+
+// ProtocolUpgradeTable is a data structure capable of specifying which protocol versions are active at which block heights
+type ProtocolUpgradeTable struct {
+	Network  string
+	upgrades []protocolUpgrade
+}
+
+// NewProtocolUpgradeTable creates a new ProtocolUpgradeTable that only tracks upgrades for the given network
+func NewProtocolUpgradeTable(network string) *ProtocolUpgradeTable {
+	return &ProtocolUpgradeTable{
+		Network: network,
+	}
+}
+
+// Add configures an upgrade for a network. If the network doesn't match the PUT's network, this upgrade will be ignored.
+func (put *ProtocolUpgradeTable) Add(network string, version uint64, effectiveAt *types.BlockHeight) {
```

Comment:
There's no need for this table to be mutable - AFAICT it will always be fully known at construction. Given that, I would much prefer it were immutable, and constructed full-formed by constructor function or builder pattern. That will allow easier enforcing of invariants (like a mandatory height-zero version) and simplify the logic of inserting into the slice down to just sorting a slice once.